### ISSUE TITLE
Weird FormattedString issue on 10.15

### ISSUE THREAD

**Issue by frankrolf** [typemytype/drawbot#281]
I installed the 10.15 Beta on my secondary computer just to poke around a bit.
When trying out some basic DrawBot, I stumbled upon this problem:

```
content = 'abc'
font_name = 'LucidaGrande'
fs = FormattedString()
fs.append(content, font=font_name, fontSize=20)
fs.append(' ')
fs.append(font_name, font='InputMono-Regular', fontSize=10)

text(fs, (100, 200))
```

The script runs, and displays output, but also the following traceback:

```
Traceback (most recent call last):
  File "<untitled>", line 8, in <module>
  File "drawBot/drawBotDrawingTools.pyc", line 1611, in text
  File "drawBot/drawBotDrawingTools.pyc", line 1779, in textBox
  File "drawBot/context/baseContext.pyc", line 2310, in clippedText
  File "drawBot/context/baseContext.pyc", line 1180, in __getitem__
  File "drawBot/context/baseContext.pyc", line 904, in __init__
  File "drawBot/context/baseContext.pyc", line 912, in _setAttribute
  File "drawBot/context/baseContext.pyc", line 1354, in openTypeFeatures
  File "drawBot/context/baseContext.pyc", line 1368, in listOpenTypeFeatures
  File "drawBot/misc.pyc", line 362, in wrapper
  File "drawBot/context/tools/openType.pyc", line 509, in getFeatureTagsForFontName
  File "drawBot/context/tools/openType.pyc", line 495, in getFeatureTagsForDescriptions
  File "objc/_convenience_mapping.pyc", line 18, in __getitem__objectForKey_
  File "objc/_convenience.pyc", line 101, in container_unwrap
KeyError: 'CTFeatureSelectorIdentifier'

```

It is entirely possible this has nothing to do with DrawBot at all, but I thought it might be interesting to know about it.

**Comment by jjgod** [comment]
For macOS 10.15, feature type dictionaries can have `kCTFontOpenTypeFeatureTag` instead of `kCTFontFeatureTypeIdentifierKey`. Likewise, feature selector dictionaries can have `kCTFontOpenTypeFeatureValue` instead of `kCTFontFeatureSelectorIdentifierKey`.

I believe for InputMono you will be seeing something like this:

        {
        CTFeatureOpenTypeTag = dnom;
        CTFeatureTypeExclusive = 1;
        CTFeatureTypeName = "'dnom'";
        CTFeatureTypeSelectors =         (
                        {
                CTFeatureOpenTypeValue = 0;
                CTFeatureSelectorDefault = 1;
                CTFeatureSelectorName = Off;
            },
                        {
                CTFeatureOpenTypeValue = 1;
                CTFeatureSelectorName = On;
            }
        );
    },
        {
        CTFeatureOpenTypeTag = numr;
        CTFeatureTypeExclusive = 1;
        CTFeatureTypeName = "'numr'";
        CTFeatureTypeSelectors =         (
                        {
                CTFeatureOpenTypeValue = 0;
                CTFeatureSelectorDefault = 1;
                CTFeatureSelectorName = Off;
            },
                        {
                CTFeatureOpenTypeValue = 1;
                CTFeatureSelectorName = On;
            }
        );
    },
        {
        CTFeatureOpenTypeTag = salt;
        CTFeatureTypeExclusive = 1;
        CTFeatureTypeName = "'salt'";
        CTFeatureTypeSelectors =         (
                        {
                CTFeatureOpenTypeValue = 0;
                CTFeatureSelectorDefault = 1;
                CTFeatureSelectorName = Off;
            },
                        {
                CTFeatureOpenTypeValue = 1;
                CTFeatureSelectorName = On;
            }
        );
    }

So in DrawBot code please don't assume `kCTFontFeatureTypeIdentifierKey ` and `kCTFontFeatureSelectorIdentifierKey` always exists.

**Comment by justvanrossum** [comment]
After some experimentation, I wonder: am I right to conclude that the `kCTFontOpenTypeFeatureTag` approach will not work in macOS 10.10?

**Comment by justvanrossum** [comment]
Although `kCTFontOpenTypeFeatureTag` has been defined since 10.10: https://developer.apple.com/documentation/coretext/kctfontopentypefeaturetag?language=objc

**Comment by jjgod** [comment]
They work since then.

**Comment by folengo** [comment]
https://gist.github.com/folengo/f1dc5cdf934f90235abf050ee3606a09

**Comment by justvanrossum** [comment]
Great, thanks @jjgod and @folengo for the example. Now we need to figure out how to get a list of supported OT feature tags from the font, so we can fix our `listOpenTypeFeatures()`.

So far, the dict returned by
`CoreText.CTFontDescriptorCopyAttribute(descriptor, CoreText.kCTFontFeaturesAttribute)`
doesn't contain any `CoreText.kCTFontOpenTypeFeatureTag` keys, so I assume we're doing something wrong there.

**Comment by justvanrossum** [comment]
I don't get the example to produce the expected results on macOS 10.10, see my comment on the gist.

**Comment by typemytype** [comment]
if `kCTFontOpenTypeFeatureTag` works nice for applying fea tags in a font while typesetting, Im fine with extracting feature tags with fontTools for `listOpenTypeFeatures()` (similar as DrawBot does for instances)

**Comment by typemytype** [comment]
thanks @folengo (I dont know why Ive made it so difficult in the first place...)

https://gist.github.com/typemytype/68bbf615649e3a4a3281709d03cee4e9
<img width="1392" alt="Screen Shot 2019-07-25 at 09 08 33" src="https://user-images.githubusercontent.com/1190358/61853270-cb091c00-aebb-11e9-92b7-97150d43e8ed.png">

next is to extract the OT feature tags, could be done with fontTools, but then drawBot is missing out the 'defaults'. The current implementation of `listOpenTypeFeatures()` returns a dict `{tag: True/False}`. Some  features are on by default, some not...

**Comment by justvanrossum** [comment]
The mechanism appears to work on 10.10, except numr/dnom are still not supported there.

**Comment by jjgod** [comment]
Yes, according to @folengo’s testing that only works since 10.13.

**Comment by jjgod** [comment]
> So far, the dict returned by
> `CoreText.CTFontDescriptorCopyAttribute(descriptor, CoreText.kCTFontFeaturesAttribute)`
> doesn't contain any `CoreText.kCTFontOpenTypeFeatureTag` keys, so I assume we're doing something wrong there.

You are right, it appears that the features can be converted to AAT type/selectors are still only returned in AAT type/selectors dict while the OpenType specific features like `dnum`/`numr` are returned in dict with `CTFeatureOpenTypeTag`.

Like you said, you probably need a different way to extract all the supported features, otherwise you will need a reserse lookup table like HarfBuzz has: https://github.com/harfbuzz/harfbuzz/blob/master/src/hb-aat-layout.cc#L92

**Comment by folengo** [comment]
@justvanrossum it may be possible to look for `CTFeatureOpenTypeTag` if [this](https://github.com/typemytype/drawbot/blob/master/drawBot/context/tools/openType.py#L493) raises a KeyError. Then inspect the `featureDescription` dict for the tag (`CTFeatureOpenTypeTag`) and its selectors' default value – which I think should always be off?

**Comment by folengo** [comment]
@typemytype The current implementation is nice in the sense that it also knows about AAT feature types and selectors that are not mapped to OT features. However I don't think access through AAT type/selectors is exposed through the DrawBot API but I'd be happy to be proven wrong.

**Comment by justvanrossum** [comment]
This should be fixed with #307, which will be included in the upcoming release. Please reopen if the problem persists.
